lib/cloud_controller/opi/apps_client.rb

```diff
@@ -92,29 +102,64 @@ def routes(process)
       routing_info = VCAP::CloudController::Diego::Protocol::RoutingInfo.new(process).routing_info
       http_routes = (routing_info['http_routes'] || []).map do |i|
         {
-          hostnames:         [i['hostname']],
+          hostname:         i['hostname'],
           port:              i['port']
         }
       end
 
       { 'cf-router' => http_routes }
     end
 
-    def recursive_ostruct(hash)
-      OpenStruct.new(hash.map { |key, value|
-        new_val = value.is_a?(Hash) ? recursive_ostruct(value) : value
-        [key, new_val]
-      }.to_h)
+    def environment_variables(process)
+      initial_env = ::VCAP::CloudController::EnvironmentVariableGroup.running.environment_json
+      opi_env = initial_env.merge(process.environment_json || {}).
+                merge('VCAP_APPLICATION' => vcap_application(process), 'MEMORY_LIMIT' => "#{process.memory}m").
+                merge(SystemEnvPresenter.new(process.service_bindings).system_env)
+
+      opi_env = opi_env.merge(DATABASE_URL: process.database_uri) if process.database_uri
+      opi_env.merge(port_environment_variables(process))
+    end
+
+    def port_environment_variables(process)
+      port = ports(process).first
+      {
+        PORT: port.to_s,
+        VCAP_APP_PORT: port.to_s,
+        VCAP_APP_HOST: '0.0.0.0'
+      }
     end
 
     def vcap_application(process)
-      process.environment_json.merge(VCAP_APPLICATION: VCAP::VarsBuilder.new(process).to_hash)
+      VCAP::VarsBuilder.new(process).to_hash.reject do |k, _v|
+        [:users].include? k
+      end
     end
 
     def process_guid(process)
       "#{process.guid}-#{process.version}"
     end
 
+    def ports(process)
+      ::VCAP::CloudController::Diego::Protocol::OpenProcessPorts.new(process).to_a
+    end
+
+    def generate_volume_mounts(process)
+      app_volume_mounts   = VCAP::CloudController::Diego::Protocol::AppVolumeMounts.new(process.app).as_json
+      proto_volume_mounts = []
+
+      app_volume_mounts.each do |volume_mount|
+        if volume_mount['device']['mount_config'].present? && volume_mount['device']['mount_config']['name'].present?
```

Comment:
From what I can tell, this test always fails as `volume_mount['device']['mount_config']` is a string (see https://github.com/cloudfoundry/cloud_controller_ng/blob/3919b14ea002ef62c7869a7900d6834cc222e118/lib/diego/bbs/models/volume_mount.pb.rb#L31 ) and so your doing `STRING['name']`. If `STRING` contains the substring anywhere, this returns `"name"` and test test succeeds (and the `volume_id` field will be set to `"name"`. If not, the test fails.

Reply:
Hi @ericpromislow .  We spent some time looking at this today.  

As we understand it the volume mounts being parsed here are the volume mounts that were received by cloud controller in response to a `bind-service` operation on a volume service broker.  If we look at the open service broker api, mount_config is [specified as an object](https://github.com/openservicebrokerapi/servicebroker/blob/master/spec.md#device-object), not a string.   These volume mounts are stored encrypted in the capi database before being used to desire an LRP.   Later on when the app is desired this info is retrieved and decrypted.  

In the case that capi is talking to diego the volume mounts are transformed into the bbs model you reference and as part of that the `mount_config` is [serialized to a string](https://github.com/cloudfoundry/cloud_controller_ng/blob/master/lib/cloud_controller/diego/app_recipe_builder.rb#L161).  

In the case that capi is talking to OPI the volume mounts are parsed down to a much simpler model of just the volume name and mount path (i.e. destination of the volume mount inside the eventual container).  For context, the eirini team didn't want the full volume mount specification as used by diego/bbs today.

Either way, the info being deserialized out of the database is a hash and therefore we believe this code is ok.  We have restested it again today and confirmed that we see correct values everywhere.    

Please let us know if you see it differently we are not experts in CAPI, or OPI/Eirini.